File: march-7th-app/src-tauri/src/build_info.rs
use serde::Serialize;

#[derive(Clone, Copy, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildInfo {
    schema_version: u8,
    app_version: &'static str,
    target: &'static str,
    source_commit: Option<&'static str>,
    source_state: &'static str,
}

const COMMIT: &str = env!("MARCH_SOURCE_COMMIT");
pub const BUILD_INFO: BuildInfo = BuildInfo {
    schema_version: 1,
    app_version: env!("CARGO_PKG_VERSION"),
    target: env!("MARCH_BUILD_TARGET"),
    source_commit: if COMMIT.is_empty() {
        None
    } else {
        Some(COMMIT)
    },
    source_state: env!("MARCH_SOURCE_STATE"),
};

#[tauri::command]
pub fn get_build_info() -> BuildInfo {
    BUILD_INFO
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn command_and_cli_value_have_one_exact_contract() {
        let value = serde_json::to_value(get_build_info()).unwrap();
        assert_eq!(get_build_info(), BUILD_INFO);
        assert_eq!(value.as_object().unwrap().len(), 5);
        assert_eq!(value["schemaVersion"], 1);
        assert_eq!(value["appVersion"], env!("CARGO_PKG_VERSION"));
        assert_eq!(value["target"], env!("MARCH_BUILD_TARGET"));
        assert!(matches!(
            value["sourceState"].as_str(),
            Some("clean" | "modified" | "unknown")
        ));
        match BUILD_INFO.source_commit {
            Some(commit) => {
                assert_eq!(commit.len(), 40);
                assert!(commit
                    .bytes()
                    .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b)));
                assert_ne!(BUILD_INFO.source_state, "unknown");
            }
            None => assert_eq!(BUILD_INFO.source_state, "unknown"),
        }
    }
}

File: march-7th-app/src/adapters/build-info.ts
import { invoke } from "@tauri-apps/api/core";

export interface BuildInfo {
  schemaVersion: 1;
  appVersion: string;
  target: string;
  sourceCommit: string | null;
  sourceState: "clean" | "modified" | "unknown";
}

export function parseBuildInfo(value: unknown): BuildInfo {
  if (!value || typeof value !== "object") throw Error("Invalid build identity");
  const info = value as Record<string, unknown>;
  if (info.schemaVersion !== 1
    || typeof info.appVersion !== "string" || info.appVersion.length > 64 || !/^\d+\.\d+\.\d+(?:[-+][\w.-]+)?$/.test(info.appVersion)
    || typeof info.target !== "string" || info.target.length > 128 || !/^[a-z0-9_]+(?:-[a-z0-9_]+)+$/.test(info.target)
    || !["clean", "modified", "unknown"].includes(info.sourceState as string)
    || (info.sourceState === "unknown" ? info.sourceCommit !== null : typeof info.sourceCommit !== "string" || !/^[a-f0-9]{40}$/.test(info.sourceCommit))) {
    throw Error("Invalid build identity");
  }
  return { schemaVersion: 1, appVersion: info.appVersion, target: info.target, sourceCommit: info.sourceCommit as string | null, sourceState: info.sourceState as BuildInfo["sourceState"] };
}

export async function readBuildInfo(): Promise<BuildInfo> {
  return parseBuildInfo(await invoke<unknown>("get_build_info"));
}
